### ISSUE TITLE
Could you please give me some advice for more class finetune from jester?

### ISSUE THREAD

**Issue by asdfqwer2015** [ahmetgunduz/Real-time-GesRec#16]
Hi, ahmetgunduz:
  I tried to finetune a model(classifier) trained on jester to my own gesture datasets, but the performance is awful. Could you please give me some advice? And Could you please share some details for your ego finetuning experiment details?(lr? freeze some layers? freeze bn?)
  My own gesture dataset has 88 classes, almost the union of jester gesture class and ego gesture class. And only a few samples per class, (train: 42 samples, val: 6 samples). The dataset is small and with distortion(wide angle camera), second person perspective, and similar with jester.
  
  In my experiment, the arch of trained model is resnet34_0.5channels (4 block layers: [3, 4, 6, 3]). Here are results:
a. only train fc layer, all conv and bn are frozen, lr 0.001, dropout 0.7, performance: train 0.573, val: 0.463
b. block layer4 and fc, conv1 and layer1~3(include bn) are frozen , lr 0.01, dropout 0.7, performance: train 0.743, val: 0.447
c. entire model, large lr 0.01 due to my bug, no dropout, but the model got best performance: train0.909, val0.582
d. train from scratch, lr 0.01, dropout 0.7, performance: train 0.712, val 0.329

I also found the pretrained model from jester may predict some `swiping` case in my own dataset as `sliding` due to distortion.

It seems the model has not benefit much from finetuning due to more classes than jester and distortion. Could you give me some advice?

Thanks.

**Comment by ahmetgunduz** [comment]
Hi,

The size of your dataset is so small. Imagine in NvGesture dataset, there is 1050  training  and  482 test videos which is considered as small dataset. 88 classes with ~ 58% validation accuracy is not bad at all. I would suggest that you gather more data. For my project, I did not freeze the model at all since EgoGesture and NvGesture datasets are different from jester dataset. The learning  rate  is  started  with  0.01, and  divided  by  10  at  10thand  25th epochs and training ended at 30th epoch. 

Another solution worth to try may be extracting feature vector using pretrained model and build a Predictive model on top of it like Random Forest, XGBoost, etc. However, I do not think that would perform better. 

Best,
Ahmet

**Comment by asdfqwer2015** [comment]
Thanks for your advice and details for EGO experiments.
I'll add more samples as I can.
I suspect that distortion may be the main reason for the model not benefit much from finetuning except for lacking of data. The pretrained model(Jester) can not work well on my own dataset for even same gesture definition. The model can classify temporal motion clearly, but confused about spatial gesture info(e.g. swiping -> sliding). And I'm trying to add more level Elastic augment on pretraining. 
Do you think distortion can lead this? If so, could you please give me any advice for handling distortion?

**Comment by asdfqwer2015** [comment]
Thanks.

**Comment by ahmetgunduz** [comment]
I see, distortion can definitely cause that. Augmentation would be one of my answers as well. Another solution may be using a better camera maybe if possible. Also are you sure that you are normalizing input correctly. I would check some statistics (mean , variance etc) for your own data set and compare those with Jester.

**Comment by asdfqwer2015** [comment]
En, I checked normalizing again, the input for both Jester and my own dataset normalized by div 255, i.e. 0-1 range. I would like to using a normal camera instead of the wide angle camera, but in my case, I've no choice. 

Do you think attention mechanism should helpful for gesture classification? I noticed:
a. I also tried few-shot learning for Jester, 2 sliding 2 swiping and 1 other gesture are chosen as novel classes. And the performance are base 91%, novel 90%, both 75%. All of base classes and novel classes are from Jester, but both_acc is not high, the model confused for gesture shape but not motion, so the model may not robust for spatial gesture shape? And lack of robustness of spatial gesture shape also difficult to generalize without enough training. p.s. https://github.com/gidariss/FewShotWithoutForgetting
b. Since the model often misclassified gesture due to gesture shape(e.g. number `1` vs number `2`, `swiping` vs `sliding`) but not motion. 
Thus, will attention mechanism handle gesture shape? How do you think of this?

Thanks for you detailed advice again.

**Comment by asdfqwer2015** [comment]
And would zero-mean normalization for this case works better?

**Comment by ahmetgunduz** [comment]
Attention may contribute but for your dataset I do not think that would affect considerably. This is because data size would still be an issue to train with attention. 

It seems models cannot learn the spatial information correctly. I have just remembered that when you fine-tune the first convolution is randomly initialized. Can you make sure to freeze all the model except the final fully connected layer and first convolution layer? That might work.

About zero-mean normalization, It is standard in image to do standardization (x-std)/mean. Zero-mean would not be better than that I presume. The important thing is you calculate the mean and std for your dataset correctly. 

Hope this helps!

**Comment by asdfqwer2015** [comment]
En, got it. Thanks again.
I'll enlarge my dataset and try zero-mean noramlization.

I've no more questions.  If you want, you could close this issue.
:)

**Comment by ahmetgunduz** [comment]
Great! have fun :)
